src/queue.rs

```diff
@@ -0,0 +1,810 @@
+    {
+        fn new(location: GuestAddress, mem: &'a GuestMemoryMmap) -> Self {
+            SomeplaceInMemory {
+                location,
+                mem,
+                phantom: PhantomData,
+            }
+        }
+
+        // Reads from the actual memory location.
+        pub fn get(&self) -> T {
+            self.mem.read_obj(self.location).unwrap()
+        }
+
+        // Writes to the actual memory location.
+        pub fn set(&self, val: T) {
+            self.mem.write_obj(val, self.location).unwrap()
+        }
+
+        // This function returns a place in memory which holds a value of type U, and starts
+        // offset bytes after the current location.
+        fn map_offset<U>(&self, offset: GuestUsize) -> SomeplaceInMemory<'a, U> {
+            SomeplaceInMemory {
+                location: self.location.checked_add(offset).unwrap(),
+                mem: self.mem,
+                phantom: PhantomData,
+            }
+        }
+
+        // This function returns a place in memory which holds a value of type U, and starts
+        // immediately after the end of self (which is location + sizeof(T)).
+        fn next_place<U>(&self) -> SomeplaceInMemory<'a, U> {
+            self.map_offset::<U>(mem::size_of::<T>() as u64)
+        }
+
+        fn end(&self) -> GuestAddress {
+            self.location
+                .checked_add(mem::size_of::<T>() as u64)
+                .unwrap()
+        }
+    }
+
+    // Represents a virtio descriptor in guest memory.
+    pub struct VirtqDesc<'a> {
+        pub addr: SomeplaceInMemory<'a, u64>,
+        pub len: SomeplaceInMemory<'a, u32>,
+        pub flags: SomeplaceInMemory<'a, u16>,
+        pub next: SomeplaceInMemory<'a, u16>,
+    }
+
+    impl<'a> VirtqDesc<'a> {
+        fn new(start: GuestAddress, mem: &'a GuestMemoryMmap) -> Self {
+            assert_eq!(start.0 & 0xf, 0);
+
+            let addr = SomeplaceInMemory::new(start, mem);
+            let len = addr.next_place();
+            let flags = len.next_place();
+            let next = flags.next_place();
+
+            VirtqDesc {
+                addr,
+                len,
+                flags,
+                next,
+            }
+        }
+
+        fn start(&self) -> GuestAddress {
+            self.addr.location
+        }
+
+        fn end(&self) -> GuestAddress {
+            self.next.end()
+        }
+
+        pub fn set(&self, addr: u64, len: u32, flags: u16, next: u16) {
+            self.addr.set(addr);
+            self.len.set(len);
+            self.flags.set(flags);
+            self.next.set(next);
+        }
+    }
+
+    // Represents a virtio queue ring. The only difference between the used and available rings,
+    // is the ring element type.
+    pub struct VirtqRing<'a, T> {
+        pub flags: SomeplaceInMemory<'a, u16>,
+        pub idx: SomeplaceInMemory<'a, u16>,
+        pub ring: Vec<SomeplaceInMemory<'a, T>>,
+        pub event: SomeplaceInMemory<'a, u16>,
```

Comment:
`flags`, `idx` and `event` can be `VolatileRef<'a, u16>`. With rust-vmm/vm-memory#19 `ring` could also become `VolatileArrayRef<'a, T>`.